File: apps/wallet-mobile/etrid-wallet/services/WalletConnectService.ts
/**
 * WalletConnect Service
 * Handles WalletConnect v2 protocol for connecting with dApps
 */

import {
  WalletConnectSession,
  WalletConnectProposal,
  TransactionRequest,
} from '@/types/dapp';

export class WalletConnectService {
  private sessions: Map<string, WalletConnectSession> = new Map();
  private proposals: Map<string, WalletConnectProposal> = new Map();
  private client: any = null;

  /**
   * Initialize WalletConnect client
   */
  async initialize(): Promise<void> {
    // TODO: Initialize WalletConnect v2 client
    // This would use @walletconnect/sign-client
    /*
    this.client = await SignClient.init({
      projectId: process.env.WALLETCONNECT_PROJECT_ID,
      metadata: {
        name: 'Ëtrid Wallet',
        description: 'Ëtrid Mobile Wallet',
        url: 'https://etrid.io',
        icons: ['https://etrid.io/icon.png']
      }
    });

    // Register event listeners
    this.client.on('session_proposal', this.handleSessionProposal);
    this.client.on('session_request', this.handleSessionRequest);
    this.client.on('session_delete', this.handleSessionDelete);
    */
  }

  /**
   * Pair with dApp using URI (from QR code or deep link)
   */
  async pair(uri: string): Promise<void> {
    try {
      // Parse WalletConnect URI
      if (!uri.startsWith('wc:')) {
        throw new Error('Invalid WalletConnect URI');
      }

      // Pair with the dApp
      // await this.client.pair({ uri });

      // The session_proposal event will be triggered
      console.log('Pairing initiated with URI:', uri);
    } catch (error) {
      console.error('Failed to pair:', error);
      throw error;
    }
  }

  /**
   * Approve session proposal
   */
  async approveSession(proposalId: string): Promise<WalletConnectSession> {
    const proposal = this.proposals.get(proposalId);
    if (!proposal) {
      throw new Error('Proposal not found');
    }

    try {
      // Get user's account and chain
      const account = await this.getCurrentAccount();
      const chainId = await this.getCurrentChainId();

      // Approve the session
      /*
      const { topic } = await this.client.approve({
        id: Number(proposalId),
        namespaces: {
          eip155: {
            accounts: [`eip155:${chainId}:${account}`],
            methods: proposal.permissions.methods,
            events: proposal.permissions.events,
          }
        }
      });
      */

      // Create session object
      const session: WalletConnectSession = {
        id: proposalId,
        topic: 'mock-topic', // Would be from client.approve()
        dAppUrl: proposal.proposer.url,
        dAppName: proposal.proposer.name,
        dAppIcon: proposal.proposer.icons[0],
        permissions: proposal.permissions.methods,
        chains: proposal.permissions.chains,
        createdAt: new Date(),
        expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000), // 7 days
      };

      this.sessions.set(session.id, session);
      this.proposals.delete(proposalId);

      // Persist to storage
      await this.saveSessions();

      return session;
    } catch (error) {
      console.error('Failed to approve session:', error);
      throw error;
    }
  }

  /**
   * Reject session proposal
   */
  async rejectSession(proposalId: string): Promise<void> {
    const proposal = this.proposals.get(proposalId);
    if (!proposal) {
      throw new Error('Proposal not found');
    }

    try {
      // Reject the session
      /*
      await this.client.reject({
        id: Number(proposalId),
        reason: {
          code: 5000,
          message: 'User rejected'
        }
      });
      */

      this.proposals.delete(proposalId);
    } catch (error) {
      console.error('Failed to reject session:', error);
      throw error;
    }
  }

  /**
   * Sign transaction for WalletConnect session
   */
  async signTransaction(
    sessionId: string,
    transaction: TransactionRequest
  ): Promise<string> {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }

    // Request user approval
    const approved = await this.requestTransactionApproval(session, transaction);
    if (!approved) {
      throw new Error('User rejected transaction');
    }

    // Sign transaction
    const signedTx = await this.signTransactionInternal(transaction);
    return signedTx;
  }

  /**
   * Sign message for WalletConnect session
   */
  async signMessage(sessionId: string, message: string): Promise<string> {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }

    // Request user approval
    const approved = await this.requestSignatureApproval(session, message);
    if (!approved) {
      throw new Error('User rejected signature');
    }

    // Sign message
    const signature = await this.signMessageInternal(message);
    return signature;
  }

  /**
   * Get all active sessions
   */
  async getSessions(): Promise<WalletConnectSession[]> {
    // Load from storage if not in memory
    if (this.sessions.size === 0) {
      await this.loadSessions();
    }

    return Array.from(this.sessions.values());
  }

  /**
   * Get pending proposals
   */
  async getProposals(): Promise<WalletConnectProposal[]> {
    return Array.from(this.proposals.values());
  }

  /**
   * Disconnect session
   */
  async disconnect(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }

    try {
      // Disconnect the session
      /*
      await this.client.disconnect({
        topic: session.topic,
        reason: {
          code: 6000,
          message: 'User disconnected'
        }
      });
      */

      this.sessions.delete(sessionId);
      await this.saveSessions();
    } catch (error) {
      console.error('Failed to disconnect session:', error);
      throw error;
    }
  }

  /**
   * Handle incoming session proposal
   */
  private handleSessionProposal = (proposal: any): void => {
    const wcProposal: WalletConnectProposal = {
      id: proposal.id.toString(),
      proposer: {
        name: proposal.params.proposer.metadata.name,
        description: proposal.params.proposer.metadata.description,
        url: proposal.params.proposer.metadata.url,
        icons: proposal.params.proposer.metadata.icons,
      },
      permissions: {
        chains: proposal.params.requiredNamespaces.eip155?.chains || [],
        methods: proposal.params.requiredNamespaces.eip155?.methods || [],
        events: proposal.params.requiredNamespaces.eip155?.events || [],
      },
      expiresAt: new Date(proposal.params.expiryTimestamp * 1000),
    };

    this.proposals.set(wcProposal.id, wcProposal);

    // Notify UI (would emit event or use callback)
    console.log('New session proposal:', wcProposal);
  };

  /**
   * Handle incoming session request (transaction, signature, etc.)
   */
  private handleSessionRequest = async (request: any): Promise<void> => {
    const { topic, params } = request;
    const session = Array.from(this.sessions.values()).find(
      (s) => s.topic === topic
    );

    if (!session) {
      console.error('Session not found for request:', topic);
      return;
    }

    const { request: sessionRequest } = params;

    try {
      let result: any;

      switch (sessionRequest.method) {
        case 'eth_sendTransaction':
          result = await this.signTransaction(
            session.id,
            sessionRequest.params[0]
          );
          break;

        case 'personal_sign':
          result = await this.signMessage(session.id, sessionRequest.params[0]);
          break;

        case 'eth_signTypedData':
        case 'eth_signTypedData_v4':
          result = await this.signTypedData(
            session.id,
            sessionRequest.params[1]
          );
          break;

        default:
          throw new Error(`Unsupported method: ${sessionRequest.method}`);
      }

      // Send response
      /*
      await this.client.respond({
        topic,
        response: {
          id: request.id,
          jsonrpc: '2.0',
          result
        }
      });
      */
    } catch (error: any) {
      // Send error response
      /*
      await this.client.respond({
        topic,
        response: {
          id: request.id,
          jsonrpc: '2.0',
          error: {
            code: -32000,
            message: error.message
          }
        }
      });
      */
    }
  };

  /**
   * Handle session deletion
   */
  private handleSessionDelete = (event: any): void => {
    const { topic } = event;
    const session = Array.from(this.sessions.values()).find(
      (s) => s.topic === topic
    );

    if (session) {
      this.sessions.delete(session.id);
      this.saveSessions();
      console.log('Session deleted:', session.id);
    }
  };

  /**
   * Sign typed data (EIP-712)
   */
  private async signTypedData(sessionId: string, typedData: string): Promise<string> {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }

    const approved = await this.requestSignatureApproval(session, typedData);
    if (!approved) {
      throw new Error('User rejected signature');
    }

    const signature = await this.signTypedDataInternal(typedData);
    return signature;
  }

  // Storage methods
  private async saveSessions(): Promise<void> {
    const sessionsArray = Array.from(this.sessions.values());
    // TODO: Persist to AsyncStorage or SecureStore
    // await AsyncStorage.setItem('wc_sessions', JSON.stringify(sessionsArray));
  }

  private async loadSessions(): Promise<void> {
    // TODO: Load from AsyncStorage or SecureStore
    // const sessionsJson = await AsyncStorage.getItem('wc_sessions');
    // if (sessionsJson) {
    //   const sessions = JSON.parse(sessionsJson);
    //   sessions.forEach((session: WalletConnectSession) => {
    //     this.sessions.set(session.id, session);
    //   });
    // }
  }

  // Mock methods that would be implemented with actual wallet integration
  private async getCurrentAccount(): Promise<string> {
    // TODO: Get from wallet service
    return '0x0000000000000000000000000000000000000000';
  }

  private async getCurrentChainId(): Promise<string> {
    // TODO: Get from wallet service
    return '1';
  }

  private async requestTransactionApproval(
    session: WalletConnectSession,
    transaction: TransactionRequest
  ): Promise<boolean> {
    // TODO: Show approval modal in UI
    return true;
  }

  private async requestSignatureApproval(
    session: WalletConnectSession,
    data: string
  ): Promise<boolean> {
    // TODO: Show approval modal in UI
    return true;
  }

  private async signTransactionInternal(transaction: TransactionRequest): Promise<string> {
    // TODO: Implement actual transaction signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }

  private async signMessageInternal(message: string): Promise<string> {
    // TODO: Implement actual message signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }

  private async signTypedDataInternal(typedData: string): Promise<string> {
    // TODO: Implement actual typed data signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }
}

export const walletConnectService = new WalletConnectService();

File: apps/EtridWalletApp/EtridWallet/Services/WalletConnectService.swift
//
//  WalletConnectService.swift
//  EtridWallet
//
//  Service for WalletConnect v2 protocol integration
//

import Foundation
import Combine

@MainActor
class WalletConnectService: ObservableObject {
    static let shared = WalletConnectService()

    @Published var sessions: [WCSession] = []
    @Published var proposals: [WCProposal] = []
    @Published var pendingRequest: WCRequest?
    @Published var connectionState: WCSessionState = .disconnected
    @Published var isInitialized = false

    private let storageKey = "wc_sessions"
    private let proposalsKey = "wc_proposals"

    // WalletConnect client would be initialized here
    // private var client: WalletConnectClient?

    private init() {
        loadSessions()
        loadProposals()
    }

    // MARK: - Initialization

    func initialize() async throws {
        guard !isInitialized else { return }

        // TODO: Initialize WalletConnect v2 client
        /*
        client = try await WalletConnectClient.initialize(
            metadata: AppMetadata(
                name: "Ëtrid Wallet",
                description: "Ëtrid Mobile Wallet",
                url: "https://etrid.io",
                icons: ["https://etrid.io/icon.png"]
            )
        )

        // Register event listeners
        client?.sessionProposalPublisher
            .sink { [weak self] proposal in
                self?.handleSessionProposal(proposal)
            }
            .store(in: &cancellables)

        client?.sessionRequestPublisher
            .sink { [weak self] request in
                self?.handleSessionRequest(request)
            }
            .store(in: &cancellables)
        */

        isInitialized = true
    }

    // MARK: - Pairing

    func pair(uri: String) async throws {
        guard uri.hasPrefix("wc:") else {
            throw WCError.invalidURI
        }

        connectionState = .connecting

        // TODO: Implement actual pairing with WalletConnect client
        // try await client?.pair(uri: uri)

        // Simulate pairing delay
        try await Task.sleep(nanoseconds: 1_000_000_000)

        connectionState = .connected
        print("Pairing initiated with URI: \(uri)")
    }

    func pairFromQRCode(_ qrCode: String) async throws {
        try await pair(uri: qrCode)
    }

    // MARK: - Session Management

    func approveSession(_ proposal: WCProposal) async throws {
        guard let proposalIndex = proposals.firstIndex(where: { $0.id == proposal.id }) else {
            throw WCError.proposalNotFound
        }

        // Get current account info
        let address = try await getCurrentAddress()
        let chainId = try await getCurrentChainId()

        // TODO: Approve session with WalletConnect client
        /*
        let namespaces = [
            "eip155": SessionNamespace(
                accounts: ["eip155:\(chainId):\(address)"],
                methods: proposal.permissions.methods,
                events: proposal.permissions.events
            )
        ]

        let session = try await client?.approve(
            proposalId: proposal.id,
            namespaces: namespaces
        )
        */

        // Create session object
        let session = WCSession(
            id: proposal.id,
            topic: "topic_" + UUID().uuidString,
            dAppUrl: proposal.proposer.url,
            dAppName: proposal.proposer.name,
            dAppIcon: proposal.proposer.iconUrl,
            dAppDescription: proposal.proposer.description,
            permissions: proposal.permissions.methods,
            chains: proposal.permissions.chains,
            createdAt: Date(),
            expiresAt: Date().addingTimeInterval(7 * 24 * 60 * 60), // 7 days
            lastUsed: Date()
        )

        sessions.append(session)
        proposals.remove(at: proposalIndex)

        saveSessions()
        saveProposals()
    }

    func rejectSession(_ proposal: WCProposal) async throws {
        guard let proposalIndex = proposals.firstIndex(where: { $0.id == proposal.id }) else {
            throw WCError.proposalNotFound
        }

        // TODO: Reject with WalletConnect client
        /*
        try await client?.reject(
            proposalId: proposal.id,
            reason: .userRejected
        )
        */

        proposals.remove(at: proposalIndex)
        saveProposals()
    }

    func disconnect(_ session: WCSession) async throws {
        guard let sessionIndex = sessions.firstIndex(where: { $0.id == session.id }) else {
            throw WCError.sessionNotFound
        }

        connectionState = .disconnecting

        // TODO: Disconnect with WalletConnect client
        /*
        try await client?.disconnect(topic: session.topic)
        */

        sessions.remove(at: sessionIndex)
        saveSessions()

        connectionState = .disconnected
    }

    func disconnectAll() async throws {
        for session in sessions {
            try? await disconnect(session)
        }
    }

    // MARK: - Request Handling

    func approveRequest(_ request: WCRequest, result: String) async throws {
        // TODO: Send response via WalletConnect client
        /*
        try await client?.respond(
            topic: request.topic,
            response: .success(result)
        )
        */

        // Update last used
        if let index = sessions.firstIndex(where: { $0.topic == request.topic }) {
            sessions[index].lastUsed = Date()
            saveSessions()
        }

        pendingRequest = nil
    }

    func rejectRequest(_ request: WCRequest, error: String) async throws {
        // TODO: Send error response via WalletConnect client
        /*
        try await client?.respond(
            topic: request.topic,
            response: .error(WCError.userRejected)
        )
        */

        pendingRequest = nil
    }

    // MARK: - Transaction Methods

    func signTransaction(_ session: WCSession, transaction: TransactionRequest) async throws -> String {
        // TODO: Integrate with CryptoManager for signing
        // This is a placeholder
        let signedTx = "0x" + String(repeating: "0", count: 130)
        return signedTx
    }

    func signMessage(_ session: WCSession, message: String) async throws -> String {
        // TODO: Integrate with CryptoManager for signing
        let signature = "0x" + String(repeating: "0", count: 130)
        return signature
    }

    func signTypedData(_ session: WCSession, typedData: String) async throws -> String {
        // TODO: Implement EIP-712 signing
        let signature = "0x" + String(repeating: "0", count: 130)
        return signature
    }

    // MARK: - Event Handlers

    private func handleSessionProposal(_ proposal: Any) {
        // Convert to WCProposal and add to proposals list
        // This would be implemented with actual WalletConnect SDK

        let wcProposal = WCProposal(
            id: UUID().uuidString,
            proposer: WCProposer(
                name: "Sample dApp",
                description: "A sample dApp",
                url: "https://example.com",
                icons: []
            ),
            permissions: WCPermissions(
                chains: ["eip155:1"],
                methods: ["eth_sendTransaction", "personal_sign"],
                events: ["chainChanged", "accountsChanged"]
            ),
            expiresAt: Date().addingTimeInterval(300)
        )

        proposals.append(wcProposal)
        saveProposals()
    }

    private func handleSessionRequest(_ request: Any) {
        // Convert to WCRequest and handle
        // This would be implemented with actual WalletConnect SDK
    }

    private func handleSessionDelete(_ topic: String) {
        sessions.removeAll { $0.topic == topic }
        saveSessions()
    }

    // MARK: - Utilities

    func getSession(byTopic topic: String) -> WCSession? {
        sessions.first { $0.topic == topic }
    }

    func getSession(byId id: String) -> WCSession? {
        sessions.first { $0.id == id }
    }

    func cleanExpiredSessions() {
        let activeSessions = sessions.filter { !$0.isExpired }
        if activeSessions.count != sessions.count {
            sessions = activeSessions
            saveSessions()
        }
    }

    func cleanExpiredProposals() {
        let activeProposals = proposals.filter { !$0.isExpired }
        if activeProposals.count != proposals.count {
            proposals = activeProposals
            saveProposals()
        }
    }

    // MARK: - Helpers

    private func getCurrentAddress() async throws -> String {
        // TODO: Get from WalletManager
        return "0x742d35Cc6634C0532925a3b844Bc454e4438f44e"
    }

    private func getCurrentChainId() async throws -> String {
        // TODO: Get from network configuration
        return "1"
    }

    // MARK: - Storage

    private func saveSessions() {
        if let encoded = try? JSONEncoder().encode(sessions) {
            UserDefaults.standard.set(encoded, forKey: storageKey)
        }
    }

    private func loadSessions() {
        if let data = UserDefaults.standard.data(forKey: storageKey),
           let decoded = try? JSONDecoder().decode([WCSession].self, from: data) {
            sessions = decoded.filter { !$0.isExpired }
        }
    }

    private func saveProposals() {
        if let encoded = try? JSONEncoder().encode(proposals) {
            UserDefaults.standard.set(encoded, forKey: proposalsKey)
        }
    }

    private func loadProposals() {
        if let data = UserDefaults.standard.data(forKey: proposalsKey),
           let decoded = try? JSONDecoder().decode([WCProposal].self, from: data) {
            proposals = decoded.filter { !$0.isExpired }
        }
    }
}
